### PR TITLE
[Feature] Support multi-stage train loop.

### PR DESCRIPTION
**Pull request by RangiLyu**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: configs/yolox/yolox_s_8x8_300e_multi-stage-loop_coco.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+        scaling_ratio_range=(0.1, 2),
+        border=(-img_scale[0] // 2, -img_scale[1] // 2)),
+    dict(
+        type='MixUp',
+        img_scale=img_scale,
+        ratio_range=(0.8, 1.6),
+        pad_val=114.0),
+    dict(type='YOLOXHSVRandomAug'),
+    dict(type='RandomFlip', prob=0.5),
+    dict(type='Resize', scale=img_scale, keep_ratio=True),
```
Comment (hhaAndroid) [inline configs/yolox/yolox_s_8x8_300e_multi-stage-loop_coco.py:94]: remove

File: configs/yolox/yolox_s_8x8_300e_multi-stage-loop_coco.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+    dict(
```
Comment (hhaAndroid) [inline configs/yolox/yolox_s_8x8_300e_multi-stage-loop_coco.py:95]: remove

File: configs/yolox/yolox_s_8x8_300e_multi-stage-loop_coco.py [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+        type='Pad',
+        pad_to_square=True,
+        pad_val=dict(img=(114.0, 114.0, 114.0))),
+    dict(
+        type='PackDetInputs',
+        meta_keys=('img_id', 'img_path', 'ori_shape', 'img_shape',
+                   'scale_factor'))
+]
+
+train_dataloader = [strong_aug_dataloader, weak_aug_dataloader]
```
Comment (hhaAndroid) [inline configs/yolox/yolox_s_8x8_300e_multi-stage-loop_coco.py:169]: Add a note to explain why it is designed this way

File: mmdet/engine/runner/loops.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+import torch
+from mmengine.runner import BaseLoop
+from torch.utils.data import DataLoader
+
+from mmdet.registry import LOOPS
+
+
+@LOOPS.register_module()
+class MultiStageTrainLoop(BaseLoop):
```
Comment (hhaAndroid) [inline mmdet/engine/runner/loops.py:12]: UT

File: mmdet/engine/runner/loops.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    def iter(self) -> int:
+        """int: Current iteration."""
+        return self._previous_iters + self._cur_loop.iter
+
+    def run(self) -> torch.nn.Module:
+        """Execute loop."""
+        for loop in self._loop_stages:
+            self._cur_loop = loop
+            model = loop.run()
+            self._previous_epochs += loop.epochs
```
Comment (Czm369) [inline mmdet/engine/runner/loops.py:71]: Whether it is necessary to add hook positions `before_train_loop` anf  `after_train_loop` ?

**Pull request closed without merging**